# Анализ базы данных сервиса для чтения книг по подписке

**Вводная информация**

Наша компания решила быть на волне тенденций и купила крупный сервис для чтения книг по подписке. 

Цель — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Задачи исследования:
- исследовать таблицы;
- выполнить задания;
- сформулировать выводы.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine 

## Исследование данных

**books**

In [5]:
# получаем сырые данные

query = '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [6]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


В таблице 1000 строк, пропусков не обнаружено.

**authors**

In [8]:
# получаем сырые данные

query = '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)

In [9]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [10]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


В таблице 636 строк, пропусков не обнаружено.

**publishers**

In [11]:
# получаем сырые данные

query = '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)

In [12]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [13]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


В таблице 340 строк, пропусков не обнаружено.

**ratings**

In [14]:
# получаем сырые данные

query = '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)

In [15]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [16]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


В таблице 6456 строк, пропусков не обнаружено.

**reviews**

In [17]:
# получаем сырые данные

query = '''
            SELECT * FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)

In [18]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [19]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


В таблице 2793 строк, пропусков не обнаружено.

### Задание 1

**Посчитаем, сколько книг вышло после 1 января 2000 года.**

In [20]:
query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date >= '2000-01-01';
'''

books_count = pd.read_sql(query, con = engine)
books_count

count
0    821

**Вывод**

После 1 января 2000 года вышла 821 книга.

### Задание 2

**Для каждой книги посчитаем количество обзоров и среднюю оценку.**

In [21]:
query = '''
SELECT b.title,
       COUNT(DISTINCT r.review_id) AS count_review, 
       AVG(rt.rating) AS avg_rating
FROM BOOKS AS b 
LEFT JOIN ratings AS rt ON rt.book_id = b.book_id 
LEFT JOIN reviews AS r ON r.book_id = rt.book_id
GROUP BY b.book_id
ORDER BY count_review DESC;
'''

rev_rat_count = pd.read_sql(query, con = engine)
rev_rat_count

title  count_review  \
0                              Twilight (Twilight  #1)             7   
1                                  Water for Elephants             6   
2                                     The Glass Castle             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    The Curious Incident of the Dog in the Night-Time             6   
..                                                 ...           ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996                            The Natural Way to Draw             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998                          Essential Tales and Poems             0   
999  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_rating  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 3 columns]

**Вывод**

Лидер среди книг по количеству обзоров - Twilight. У книги - 7 обзоров.

### Задание 3

**Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.**

In [22]:
query = '''
SELECT p.publisher,
       COUNT(b.book_id) AS count_book
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY count_book DESC
LIMIT 1;
'''

name_publisher = pd.io.sql.read_sql(query, con = engine)
name_publisher

publisher  count_book
0  Penguin Books          42

**Вывод**

Больше всего книг выпустило издательство Penguin Books (42 книги).

### Задание 4

**Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками.**

In [23]:
query = '''
SELECT a.author,
       AVG(rt.rating) AS avg_rating
FROM authors AS a
INNER JOIN books AS b ON a.author_id = b.author_id
INNER JOIN ratings AS rt ON b.book_id = rt.book_id
WHERE b.book_id IN
    (SELECT b.book_id
     FROM books AS b
     JOIN ratings AS rt ON b.book_id = rt.book_id
     GROUP BY b.book_id
     HAVING COUNT(rt.rating_id) >= 50)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1;
'''

author_high = pd.io.sql.read_sql(query, con = engine)
author_high

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

**Вывод**

Автор с самой высокой средней оценкой книг - J.K.Rowling/Mary GrandPre.

### Задание 5

**Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.**

In [24]:
query = '''  
SELECT AVG(r.count_review) AS avg_review
FROM 
    (SELECT COUNT(review_id) AS count_review
     FROM reviews
     WHERE username IN 
                   (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 48)
     GROUP BY username) AS r;
'''
avg_count_rev = pd.io.sql.read_sql(query, con = engine)
avg_count_rev

avg_review
0        24.0

**Вывод**

Пользователи, оставившие более 48 оценок, в среднем написали - 24 обзора.

## Общий вывод

Цель — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Выполненные задачи:
- исследовать таблицы;
- выполнить задания;
- сформулировать выводы.

Было выявлено:
- После 1 января 2000 года вышла 821 книга.
- Лидер среди книг по количеству обзоров - Twilight. У книги - 7 обзоров.
- Больше всего книг выпустило издательство Penguin Books (42 книги).
- Автор с самой высокой средней оценкой книг - J.K.Rowling/Mary GrandPre.
- Самые активные пользователи (оставившие более 48 оценок) в среднем написали - 24 обзора.

**Рекомендации:**
- Стоит вкладываться в книги, которые лидируют по количеству обзоров, то есть анализировать тенденции и обсуждения на форумах любителей книг и выбирать для сервиса наиболее обсуждаемые книги. Например, книги серии "Сумерки" - если настолько обсуждаема первая книга, то остальные книги также станут обсуждаемыми.
- Стоит обратить внимание на издательство Penguin Books, поскольку оно лидирует по выпуску книг.
- Стоит обратить внимание на автора J.K.Rowling и следить за выпуском новых книг, поскольку данный автор имеет высокую среднюю оценку, а значит остальные книги читатели будут также стремиться оценить.